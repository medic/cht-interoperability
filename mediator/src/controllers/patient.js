const axios = require('axios');
const { genereateFHIRPatientResource } = require('../utils/patient');
const FHIR_ROOT_URL = 'http://openhim-core:5001/fhir';

async function createPatient(CHTpatientDoc) {
  const FHITPatientResource = genereateFHIRPatientResource(CHTpatientDoc);

  try {
    const res = await axios.post(`${FHIR_ROOT_URL}/Patient`, FHITPatientResource, { auth: {
      username: 'interop-client',
      password: 'interop-password'
    }});
    return {status: res.status, patient: res.data}
  } catch (error) {
    console.error(error);
    return {status: error.status, patient: error.data}
  }
}

module.exports = {
  createPatient
};
