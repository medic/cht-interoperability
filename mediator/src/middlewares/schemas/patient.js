const joi = require('joi');

const createPatientSchema = joi.object({
  name: joi.string().optional(),
  _id: joi.string().optional(),
  sex: joi.string().optional(),
  birthDate: joi.string().optional(),
});

module.exports = {
  createPatientSchema,
};
