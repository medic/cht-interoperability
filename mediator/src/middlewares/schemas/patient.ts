import joi from "joi";


export const createPatientSchema = joi.object({
  parent: joi.any().optional(),
  type: joi.string().optional(),
  name: joi.string().required(),
  _id: joi.string().required(),
  id: joi.string().optional(),
  sex: joi.string().trim().valid('male', 'female', 'other', 'unkown').required(),
  // validate expecting YYYY-MM-DD
  date_of_birth: joi
    .string()
    .regex(/((?:19|20)\d\d)-(0?[1-9]|1[012])-([12][0-9]|3[01]|0?[1-9])/)
    .required()
    .messages({
      "object.regex": "Invalid date expecting YYYY-MM-DD",
      "string.pattern.base": "Invalid date expecting YYYY-MM-DD"
    })
});
