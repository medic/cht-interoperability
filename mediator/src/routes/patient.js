const {Router} = require('express');
const router = Router();
const patientController = require('../controllers/patient');
const {validateBodyAgainst} = require('../middlewares');
const {createPatientSchema} = require('../middlewares/schemas/patient');

router.post('/',
  validateBodyAgainst(createPatientSchema),
  async function(req, res) {
    const {status, patient} = await patientController.createPatient(req.body);
    res.send({status, patient});
});

module.exports = router;
