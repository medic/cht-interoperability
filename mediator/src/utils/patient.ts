function genereateFHIRPatientResource(patient: any) {
  const patientLastName = patient.name.split(' ').slice(-1);
  const FHITPatientResource = {
    resourceType: 'Patient',
    id: patient.id,
    identifier: [
      {
        system: 'cht',
        value: patient._id
      }
    ],
    name: [
      {
        use: 'official',
        family: patientLastName,
        given: [patient.name]
      }
    ],
    gender: patient.sex,
    birthDate: patient.date_of_birth
  };

  return FHITPatientResource;
}

module.exports = {
  genereateFHIRPatientResource,
};
