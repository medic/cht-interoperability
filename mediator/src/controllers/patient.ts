import axios from 'axios';
import {generateFHIRPatientResource} from '../utils/patient';
import {FHIR} from '../../config';
import {logger} from '../../logger';
const {url: fhirUrl, username: fhirUsername, password: fhirPassword} = FHIR;

export async function createPatient(CHTpatientDoc: any) {
  try {
    const FHIRPatientResource = generateFHIRPatientResource(CHTpatientDoc);
    const res = await axios.post(`${fhirUrl}/Patient`, FHIRPatientResource, {auth: {
      username: fhirUsername,
      password: fhirPassword,
    }});
    return {status: res.status, patient: res.data};
  } catch (error: any) {
    logger.error(error);

    if (!error.status) {
      return {status: 400, patient: {message: error.message}};
    }

    return {status: error.status, patient: error.data};
  }
}
