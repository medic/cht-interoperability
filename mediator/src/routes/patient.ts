import { Router } from 'express';
import { validateBodyAgainst } from '../middlewares';
import { createFhirResource, validateFhirResource } from '../utils/fhir';
import { PatientSchema } from '../middlewares/schemas/patient';
import { requestHandler } from '../utils/request';

const router = Router();

const resourceType = 'Patient';

router.post(
  '/',
  validateBodyAgainst(validateFhirResource(resourceType), PatientSchema),
  requestHandler((req) => createFhirResource({ ...req.body, resourceType }))
);

export default router;
