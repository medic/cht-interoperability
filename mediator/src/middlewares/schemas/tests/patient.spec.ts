import { createPatientSchema } from "../patient";

describe("createPatientSchema", () => {
  it("accepts when all values are present", async () => {
    const data = {
      name: "John Doe",
      _id: "JOHN_DOE_ID",
      id: "OPTIONAL",
      sex: "male",
      date_of_birth: "2000-01-01",
      parent: "OPTIONAL",
      type: "OPTIONAL",
    };

    const res = await createPatientSchema.validateAsync(data);
    expect(res).toMatchSnapshot();
  });

  it("accepts when only required values are present", async () => {
    const data = {
      name: "John Doe",
      _id: "JOHN_DOE_ID",
      sex: "male",
      date_of_birth: "2000-01-01",
    };

    const res = await createPatientSchema.validateAsync(data);
    expect(res).toMatchSnapshot();
  });

  it("accepts only valid right genders", async () => {
    const data = {
      name: "John Doe",
      _id: "JOHN_DOE_ID",
      id: "OPTIONAL",
      sex: "dangled beef",
      date_of_birth: "2000-01-01",
      parent: "OPTIONAL",
      type: "OPTIONAL",
    };

    expect(createPatientSchema.validateAsync(data)).rejects.toMatchSnapshot();

    expect(
      await createPatientSchema.validateAsync({ ...data, sex: "female" })
    ).toMatchSnapshot();
    expect(
      await createPatientSchema.validateAsync({ ...data, sex: "unknown" })
    ).toMatchSnapshot();
    expect(
      await createPatientSchema.validateAsync({ ...data, sex: "other" })
    ).toMatchSnapshot();
  });

  it("accepts only dates in the format of YYYY-MM-DD", () => {
    const data = {
      name: "John Doe",
      _id: "JOHN_DOE_ID",
      id: "OPTIONAL",
      sex: "male",
      date_of_birth: "01-2000-01",
      parent: "OPTIONAL",
      type: "OPTIONAL",
    };

    expect(
      createPatientSchema.validateAsync(data)
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(
      createPatientSchema.validateAsync({
        ...data,
        date_of_birth: "01-01-2000",
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
