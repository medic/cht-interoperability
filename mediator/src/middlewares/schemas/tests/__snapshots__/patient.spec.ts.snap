// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPatientSchema accepts only dates in the format of YYYY-MM-DD 1`] = `"Invalid date expecting YYYY-MM-DD"`;

exports[`createPatientSchema accepts only dates in the format of YYYY-MM-DD 2`] = `"Invalid date expecting YYYY-MM-DD"`;

exports[`createPatientSchema accepts only valid right genders 1`] = `[ValidationError: "sex" must be one of [male, female, other, unknown]]`;

exports[`createPatientSchema accepts only valid right genders 2`] = `
{
  "_id": "JOHN_DOE_ID",
  "date_of_birth": "2000-01-01",
  "id": "OPTIONAL",
  "name": "John Doe",
  "parent": "OPTIONAL",
  "sex": "female",
  "type": "OPTIONAL",
}
`;

exports[`createPatientSchema accepts only valid right genders 3`] = `
{
  "_id": "JOHN_DOE_ID",
  "date_of_birth": "2000-01-01",
  "id": "OPTIONAL",
  "name": "John Doe",
  "parent": "OPTIONAL",
  "sex": "unknown",
  "type": "OPTIONAL",
}
`;

exports[`createPatientSchema accepts only valid right genders 4`] = `
{
  "_id": "JOHN_DOE_ID",
  "date_of_birth": "2000-01-01",
  "id": "OPTIONAL",
  "name": "John Doe",
  "parent": "OPTIONAL",
  "sex": "other",
  "type": "OPTIONAL",
}
`;

exports[`createPatientSchema accepts when all values are present 1`] = `
{
  "_id": "JOHN_DOE_ID",
  "date_of_birth": "2000-01-01",
  "id": "OPTIONAL",
  "name": "John Doe",
  "parent": "OPTIONAL",
  "sex": "male",
  "type": "OPTIONAL",
}
`;

exports[`createPatientSchema accepts when only required values are present 1`] = `
{
  "_id": "JOHN_DOE_ID",
  "date_of_birth": "2000-01-01",
  "name": "John Doe",
  "sex": "male",
}
`;
